core/offchain-worker-executor/src/executor.rs

```diff
@@ -0,0 +1,198 @@
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+use crate::error::Result;
+use ita_stf::hash::TrustedOperationOrHash;
+use itc_parentchain_light_client::{
+	concurrent_access::ValidatorAccess, BlockNumberOps, ExtrinsicSender, LightClientState,
+	NumberFor,
+};
+use itp_extrinsics_factory::CreateExtrinsics;
+use itp_stf_executor::{traits::StateUpdateProposer, ExecutedOperation};
+use itp_stf_state_handler::{handle_state::HandleState, query_shard_state::QueryShardState};
+use itp_top_pool_author::traits::{AuthorApi, OnBlockImported, SendState};
+use itp_types::{OpaqueCall, ShardIdentifier, H256};
+use log::*;
+use sp_runtime::traits::Block;
+use std::{marker::PhantomData, sync::Arc, time::Duration, vec, vec::Vec};
+
+/// Off-chain worker executor implementation.
+///
+/// Executes calls found in the top-pool and immediately applies the corresponding state diffs.
+/// - Sends confirmations for all executed calls (TODO)
+/// - Sends extrinsics for any parentchain effects (such as unshield calls).
+///
+/// The trigger to start executing calls is given when the parentchain block imported event is
+/// signaled (event listener).
+pub struct Executor<
+	ParentchainBlock,
+	TopPoolAuthor,
+	StfExecutor,
+	StateHandler,
+	ValidatorAccessor,
+	ExtrinsicsFactory,
+> {
+	top_pool_author: Arc<TopPoolAuthor>,
+	stf_executor: Arc<StfExecutor>,
+	state_handler: Arc<StateHandler>,
+	validator_accessor: Arc<ValidatorAccessor>,
+	extrinsics_factory: Arc<ExtrinsicsFactory>,
+	_phantom: PhantomData<ParentchainBlock>,
+}
+
+impl<
+		ParentchainBlock,
+		TopPoolAuthor,
+		StfExecutor,
+		StateHandler,
+		ValidatorAccessor,
+		ExtrinsicsFactory,
+	>
+	Executor<
+		ParentchainBlock,
+		TopPoolAuthor,
+		StfExecutor,
+		StateHandler,
+		ValidatorAccessor,
+		ExtrinsicsFactory,
+	> where
+	ParentchainBlock: Block<Hash = H256>,
+	StfExecutor: StateUpdateProposer,
+	TopPoolAuthor: AuthorApi<H256, ParentchainBlock::Hash>
+		+ OnBlockImported<Hash = ParentchainBlock::Hash>
+		+ SendState<Hash = ParentchainBlock::Hash>,
+	StateHandler: QueryShardState + HandleState<StateT = StfExecutor::Externalities>,
+	ValidatorAccessor: ValidatorAccess<ParentchainBlock> + Send + Sync + 'static,
+	ExtrinsicsFactory: CreateExtrinsics,
+	NumberFor<ParentchainBlock>: BlockNumberOps,
+{
+	pub fn new(
+		top_pool_author: Arc<TopPoolAuthor>,
+		stf_executor: Arc<StfExecutor>,
+		state_handler: Arc<StateHandler>,
+		validator_accessor: Arc<ValidatorAccessor>,
+		extrinsics_factory: Arc<ExtrinsicsFactory>,
+	) -> Self {
+		Self {
+			top_pool_author,
+			stf_executor,
+			state_handler,
+			validator_accessor,
+			extrinsics_factory,
+			_phantom: Default::default(),
+		}
+	}
+
+	pub fn execute(&self) -> Result<()> {
```

Comment:
In case of the sidechain, the 'slot proposer' does the work that the offchain worker executor does here. So the API and usage/client is very different. A common trait in its current state does not make sense IMO.